### PR TITLE
Added custom task queues to the rmq thread comm

### PR DESCRIPTION
**Pull request by muhrin**
Summary:

  * Added custom task queues via comm.task_queue('my_new_queue')
  * Supported in both coroutine and thread RMQ communicators
  * new dependency: pytray. Responsible for translation of thread->async calls.
  * Corresponding tests added
  * Interface *hasn't*  been exposed at kiwipy.Communicator level

Details:

For now, this isn't exposed at the `kiwipy.Communicator` level.  This
will be added later but for now it works well with kiwi and it proved a
good testing ground for developing the feature.

It works like this:

  task_queue = comm.task_queue('my_task_queue')
  task_queue.add_task_subscriber(lambda _, task: print(task))
  task_queue.task_send('Do this!')

Essentially every communicator has a default queue (which is the one
referred to by doing comm.add_task_subscriber/task_send) but users are
free to create additional ones using the above syntax.

Async versions also exist and can be iterated like this:

await for task in task_queue:
  task.process().set_result('done')

The threaded version also suppoorts this without the await syntax.

PyTray provides a LoopScheduler which does all the negotiation between
the thread running the asyncio event loop and the coroutines themselves.

task_get raises kiwipy.QueueEmpty exception if empty.

### CODE REVIEW

File: kiwipy/rmq/communicator.py [hunk 1]
```diff
@@ -401,362 +394,27 @@ async def broadcast_send(self, body, sender=None, subject=None, correlation_id=N
...
-                                         task_queue=task_queue,
-                                         task_prefetch_size=task_prefetch_size,
-                                         task_prefetch_count=task_prefetch_count,
-                                         encoder=encoder,
-                                         decoder=decoder,
-                                         testing_mode=testing_mode)
+    async def task_queue(self,
+                         queue_name: str,
+                         prefetch_size=defaults.TASK_PREFETCH_SIZE,
+                         prefetch_count=defaults.TASK_PREFETCH_COUNT) -> tasks.RmqTaskQueue:
```
Comment (sphuber) [inline kiwipy/rmq/communicator.py:408]: Are you on purpose not allowing to define different encoder/decoder and or the testing mode on a per task-queue basis?

File: kiwipy/rmq/tasks.py [hunk 2]
```diff
@@ -99,62 +138,38 @@ async def _create_task_queue(self):
...
-
-                if handled:
-                    message.ack()
-                    if not task_body.no_reply:
-                        await self._send_response(reply_body, message)
-                    break  # Done, do break the loop
-
-            if not handled:
-                # No one handled the task
-                message.reject(requeue=True)
```
[the remark this answers is not in the document]
Comment (muhrin) [reply kiwipy/rmq/tasks.py:157]: This is handled by the context manage in a more safe way, if you don't set an outcome in the context then it is automatically requeued here:
https://github.com/aiidateam/kiwipy/pull/43/files/a99e76b0e41c2859fec1253c1f4667e45e5d73fe#diff-bcf4355c138e3aeeea7ea12d5aa848ecR265

File: kiwipy/rmq/tasks.py [hunk 3]
```diff
@@ -178,14 +193,122 @@ async def _send_response(self, msg_body, incoming_message):
...
+                self._subscriber.loop().create_task(self._subscriber._send_response(reply_body, self._message))
+
+        # Clean up
+        self._finalise()
+
+    def _outcome_destroyed(self, outcome_ref):
+        # This only happens if someone called self.process() and then let the future
+        # get destroyed without setting an outcome
+        assert outcome_ref is self._outcome_ref
+        # This task will no be processed
```
[the remark this answers is not in the document]
Comment (muhrin) [reply kiwipy/rmq/tasks.py]: Haha, you don't allow lowland Scots??
It's been poshified.

File: kiwipy/rmq/messages.py [hunk 4]
```diff
@@ -55,14 +56,17 @@ def __init__(self, connection, exchange_name=defaults.MESSAGE_EXCHANGE, exchange
         self._exchange_params = exchange_params
         self._loop = self._connection.loop
 
-        self._channel = None  # type: aio_pika.Channel
-        self._exchange = None  # type: aio_pika.Exchange
+        self._channel = None  # type: typing.Optional[aio_pika.Channel]
```
[the remark this answers is not in the document]
Comment (muhrin) [reply kiwipy/rmq/messages.py:59]: In 3.5 yes.  I'd originally had them in the more modern style:

`self._channel: typing.Optional[aio_pika.Channel] = None`

Alas, 3.5 says no.

File: kiwipy/rmq/communicator.py (hunk 1 above)
Comment (muhrin) [reply kiwipy/rmq/communicator.py:408]: Yes.  I consider being in testing more and the encoding scheme to be a per communicator setting (unless the user wants to go lower level and access the `rmq` package directly)

**Comment by muhrin** [comment]
I've pushed a new commit which is to be squashed before being committed to dev.

**Comment by muhrin** [comment]
Looks like this PR caught a mild case of bonitis because of a `pytest-asyncio` release yesterday so I've had to limit the version.

**Comment by muhrin** [comment]
Ok.  Tests passing again.  Back over to you my friend.

**Pull request merged**